File: handover.py
# -*- coding: utf-8 -*-

from monta_grafo import Grafo
import numpy as np

class handover:
    def __init__(self, caminho):
        self._caminho = caminho
        self.grafo_instancia = Grafo()
        self.num_arestas = 0
        #faz a leitura do arquivo da instancia
        with open(caminho) as instancia:
            #anda pelas linhas
            i=0
            for line in instancia.readlines():
                print("line:{}\n".format(line))
                ls = line.rstrip()
                print("ls:{}\n".format(ls))
                if i==0:
                    # Lendo a primeira linha
                    linha1 = ls
                    self.numero_elementos = int(linha1)
                elif i==1:
                    self.numero_clusters = int(ls)
                elif i==2:
                    #inicializando restrição superior
                    self.U = float(ls)
                elif i>2 and i<=self.numero_elementos+2:
                    self.grafo_instancia.add_no(i-3,float(ls))
                else:
                    #calculando restrição inferior
                    #média dos pesos de todos nós
                    self.L = sum(self.grafo_instancia.pesos_nos)/self.numero_elementos
                    #lendo o array de pesos da aresta,uma linha únic
                    array_pesos = np.array([float(peso) for peso in ls.split(" ")])
                    #matrix de interligação entre os nós, peso das arestas
                    matrix_pesos_aresta = np.split(array_pesos,int(self.numero_elementos))
                    for no_ind,no_linha in enumerate(matrix_pesos_aresta):
                        for viz_ind,valor_viz in enumerate(no_linha):
                            #0.0 é quando não tem ligação
                            if valor_viz != 0.0:
                                self.grafo_instancia.add_aresta(no_ind,viz_ind,valor_viz)
                                self.num_arestas+=1


                i+=1

    def getGrafo(self):
        return self.grafo_instancia

    def get_L(self):
        return self.L

    def get_U(self):
        return self.U

    def __str__(self) -> str:
        return '--Instância--\nnumero_elementos: {}\nnumero_clusters: {}\nnum_arestas: {}\nL: {}\nU: {}'.format(
            self.numero_elementos, self.numero_clusters, self.num_arestas, self.L, self.U
        )


inst = handover('instancias/handover/20_5_270001.txt')

print(inst)

File: handover.js
const Grafo = require('./grafo')
const fs = require('fs')

/**
 * Classe onde é realizada a leitura das instâncias Handover
 */
class Handover {

    numArestas = 0
    numeroElementos = 0
    numeroClusters = 0

    /**
     * Particiona um array em uma matriz, com o tamanho da linha sendo dado
     * pelo parâmetro size
     * @param {array} array 
     * @param {number} size 
     * @returns {array[]} Matriz resultante
     */
    npSplit(array, size) {
        return array.reduce(function (rows, key, index) {
            return (index % size == 0 ? rows.push([key])
                : rows[rows.length - 1].push(key)) && rows;
        }, []);
    }

    /**
     * Construtor da classe, onde é fornecido o caminho da instância
     * @param {string} caminho 
     */
    constructor(caminho) {

        this.grafoInstancia = new Grafo()

        //faz a leitura do arquivo da instancia
        let dataBuffer = fs.readFileSync(caminho);
        const instancia = dataBuffer.toString().split('\n');


        //anda pelas linhas
        let i = 0;
        for (const line of instancia) {

            const ls = line.trim()

            if (i === 0) {
                // Lendo a primeira linha
                let linha1 = ls
                this.numeroElementos = parseInt(linha1)
            } else if (i === 1) {
                this.numeroClusters = parseInt(ls)
            } else if (i === 2) {
                //inicializando restrição superior
                this.U = parseFloat(ls)
            } else if (i > 2 && i <= this.numeroElementos + 2) {
                this.grafoInstancia.addNo(i - 3, parseFloat(ls))
            } else {
                //lendo o array de pesos da aresta,uma linha única
                let array_pesos = ls.split(" ").map(peso => parseFloat(peso))

                // Matriz de interligação entre os nós, peso das arestas
                let matrix_pesos_aresta = this.npSplit(array_pesos, parseInt(this.numeroElementos))

                // Percorre a matriz de interligação para obter as arestas
                for (const [no_ind, no_linha] of matrix_pesos_aresta.entries()) {
                    for (const [viz_ind, valor_viz] of no_linha.entries()) {
                        // Se a aresta tem peso 0, ela não vai influenciar no resultado final
                        // então, ela não é incluida no grafo final da instância
                        if (valor_viz !== 0) {
                            this.grafoInstancia.addAresta(no_ind, viz_ind, valor_viz)
                            this.numArestas += 1
                        }
                    }
                }
            }
            i += 1
        }
    }

    getNumClusters() {
        return this.numeroClusters
    }

    getNumElementos() {
        return this.numeroElementos
    }

    getGrafo() {
        return this.grafoInstancia
    }

    get_L() {
        return 0
    }

    get_U() {
        return this.U
    }

}

module.exports = Handover;